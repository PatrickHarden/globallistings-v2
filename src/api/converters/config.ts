import { mapObjIndexed } from 'ramda';
import { Config } from '../../types/config';
import flattenObject from '../../utils/flattenObj';

const replaceIntlChars = mapObjIndexed((_, key, obj) =>
  obj[key].replace(/%\(/g, '{').replace(/\)s/g, '}')
);

const convertToIntl = (messages: object) => replaceIntlChars(messages);

// TODO this should leave outside API folder as this is not an API layer anymore and just a transformation layer
export const convert = (config: Config): Config => ({
  ...config,
  i18n: convertToIntl(flattenObject(config.i18n)),
});
