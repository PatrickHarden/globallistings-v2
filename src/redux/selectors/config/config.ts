import { createSelector } from 'reselect';
import { configSelector, configMetaSelector } from '../state';

export const configStatusSelector = createSelector(
  configMetaSelector,
  config => config.status
);

export const getFilters = createSelector(
  configSelector,
  config => config.filters
);

export const siteTypeSelector = createSelector(
  configSelector,
  config => config.siteType
);
export const searchModeSelector = createSelector(
  configSelector,
  config => config.searchMode
);

// SearchConfig
export const searchConfigSelector = createSelector(
  configSelector,
  config => config.searchConfig
);

export const searchBiasSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchBias
);

export const searchRadiusRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.biasRadius
);

export const searchCountryRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.restrictToCountry
);

export const polygonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.polygons
);

export const searchTypesRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPlaceTypes
);

export const assumeZipCodeSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.allowAssumeZipCode
);

export const searchPathSortAlphabeticalSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPathSortAlphabetical
);

export const searchPathSelectorSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPathSelector
);

export const propertyIdPathSelectorSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.propertyIdSearchPathSelector
);

export const hideBuyButtonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.hideSearchToBuy
);

export const hideLetButtonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.hideSearchToLet
);

export const getSearchHeader = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchHeader
);
export const searchResultsPageSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchResultsPage
);

export const searchSuggestDebounceSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchSuggestDebounce
);

export const defaultSearchLocationNameSelector = createSelector(
  configSelector,
  config => config.searchLocationName
);

export const paramsSelector = createSelector(
  configSelector,
  config => config.params
);

export const restrictionsSelector = createSelector(
  configSelector,
  config => ({
    country: config.searchConfig.restrictToCountry,
  })
);

export const otherConfigsSelector = createSelector(
  configSelector,
  config => (config.otherConfigs ? config.otherConfigs : [])
);

export const currentConfigSelector = createSelector(
  configSelector,
  config => config.currentConfig
);
