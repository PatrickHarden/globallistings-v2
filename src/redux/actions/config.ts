import axios from 'axios';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators-singular';
import { Config } from '../../types/config';
import { State } from '../../types/state';
import { configEndpointForCurrentUsageTypeSelector } from '../selectors/plp';

export const createFetchConfigRequest = createRequestPending('FETCH_CONFIG_REQUEST');
export const createFetchConfigFailure = createRequestFailure('FETCH_CONFIG_FAILURE');
export const createFetchConfigSuccess = createRequestSuccess<Config, 'FETCH_CONFIG_SUCCESS'>(
  'FETCH_CONFIG_SUCCESS'
);

// This will be used in the future to switch between multiple configs
export const fetchConfig = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const configEndpoint = configEndpointForCurrentUsageTypeSelector(state);

  dispatch(createFetchConfigRequest());
  return axios
    .get(configEndpoint)
    .then(res => {
      dispatch(createFetchConfigSuccess(res.data));
    })
    .catch(ex => {
      dispatch(createFetchConfigFailure(ex));
    });
};
