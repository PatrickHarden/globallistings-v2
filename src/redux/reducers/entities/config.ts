import { convert } from '../../../api/converters/config';
import { createFetchConfigSuccess } from '../../actions/config';
import { ActionType } from 'typesafe-actions';
import { DeepPartial } from 'redux';
import { Config } from '../../../types/config';

export type Actions = ActionType<typeof createFetchConfigSuccess>;

export default (state: DeepPartial<Config> = {}, action: Actions): Config | {} => {
  switch (action.type) {
    case 'FETCH_CONFIG_SUCCESS':
      return convert(action.payload);
    default:
      return state;
  }
};
