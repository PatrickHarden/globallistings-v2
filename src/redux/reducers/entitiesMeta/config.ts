import * as actions from '../../actions/config';
import { ActionType } from 'typesafe-actions';
import { handleError, handleSuccess, handlePending } from './shared/meta-reducers';

export type Actions = ActionType<typeof actions>;

export default (state = {}, action: Actions) => {
  switch (action.type) {
    case 'FETCH_CONFIG_REQUEST':
      return handlePending();
    case 'FETCH_CONFIG_SUCCESS':
      return handleSuccess(action);
    case 'FETCH_CONFIG_FAILURE':
      return handleError(action);
    default:
      return state;
  }
};
