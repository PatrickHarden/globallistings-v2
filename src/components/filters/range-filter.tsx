import React, { FC } from 'react';
import styled from 'styled-components';

import { FilterOption, FilterModifier } from '../../types/config';
import { RangeFilterInfo } from '../../redux/selectors/config/filters';

import Select from '../select/select';

const Filter = styled(Select)``;

const FilterLabel = styled.label`
  flex: 1 100%;
`;

interface ContainerProps {
  readonly inline: boolean;
}
const Container = styled.div<ContainerProps>`
  display: flex;
  flex-flow: row wrap;
  justify-content: space-between;
  width: ${({ inline }) => (inline ? 'auto' : '100%')};

  ${Filter} {
    margin-left: ${({ inline }) => (inline ? '-1px' : '0')};
    width: calc(
      50%
        ${({ theme, inline }) =>
          inline ? '+ 1px' : `- ${parseInt(theme.gutter.m, undefined) / 2}em`}
    );
  }
`;

interface Props {
  readonly label?: string;
  readonly displayLabel: boolean;
  readonly filter: RangeFilterInfo;
  readonly handleChange: (key: string) => (val: string, modifier: FilterModifier) => void;
  readonly inline: boolean;
}

export const RangeFilter: FC<Props> = ({ displayLabel, filter, handleChange, inline }) => {
  const handleSelect = (item: FilterOption) =>
    handleChange(`${filter.name}_${item.type}`)(item.value, filter.modifier);

  return (
    <Container inline={inline}>
      {displayLabel && <FilterLabel>{filter.label}</FilterLabel>}
      <Filter
        {...filter.min}
        handleSelect={handleSelect}
        defaultToFirstValue={true}
        textEditable={false}
      />
      <Filter
        {...filter.max}
        handleSelect={handleSelect}
        defaultToFirstValue={true}
        textEditable={false}
      />
    </Container>
  );
};
