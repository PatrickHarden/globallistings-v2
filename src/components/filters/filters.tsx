import React, { FC } from 'react';
import styled from 'styled-components';
import { FilterInfo } from '../../redux/selectors/config/filters';
import { RangeFilter } from './range-filter';
import { SelectFilter } from './select-filter';
import { FilterModifier } from '../../types/config';

const Wrapper = styled.div``;

interface Props {
  readonly displayLabels: boolean;
  readonly filters: FilterInfo[];
  readonly handleChange: (key: string) => (val: string, modifier: FilterModifier) => void;
  readonly inline: boolean;
  readonly placement?: string;
  readonly className?: string;
}

const filterToKey = (filter: FilterInfo) => `${filter.type}-${filter.name}`;

export const Filters: FC<Props> = ({
  className,
  displayLabels,
  filters,
  handleChange,
  inline,
  placement,
}) => (
  <Wrapper className={className}>
    {filters
      .filter(filter => (placement ? filter.placement === placement : true))
      .map(filter => {
        switch (filter.type) {
          case 'select':
            return (
              <SelectFilter
                displayLabel={displayLabels}
                handleSelect={handleChange(filter.name)}
                inline={inline}
                filter={filter}
                key={filterToKey(filter)}
              />
            );
          case 'range':
            return (
              <RangeFilter
                displayLabel={displayLabels}
                filter={filter}
                key={filterToKey(filter)}
                handleChange={handleChange}
                inline={inline}
              />
            );
          default:
            console.error(`Unsupported filter type: ${filter.type}`);
            return null;
        }
      })}
  </Wrapper>
);
