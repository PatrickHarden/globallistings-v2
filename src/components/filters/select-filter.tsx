import React, { FC } from 'react';
import styled from 'styled-components';

import { SelectFilterInfo } from '../../redux/selectors/config/filters';
import { FilterModifier } from '../../types/config';

import Select from '../select/select';

interface FilterProps {
  readonly inline: boolean;
}
const Filter = styled(Select)<FilterProps>`
  margin-left: ${({ inline }) => (inline ? '-1px' : '0')};
`;

interface Props {
  readonly displayLabel: boolean;
  readonly filter: SelectFilterInfo;
  readonly handleSelect: (val: string, modifier: FilterModifier) => void;
  readonly inline: boolean;
}

export const SelectFilter: FC<Props> = ({ displayLabel, filter, handleSelect, inline }) => (
  <Filter
    {...filter}
    defaultToFirstValue={true}
    handleSelect={fo => handleSelect(fo.value, filter.modifier)}
    inline={inline}
    textEditable={false}
    labelText={displayLabel ? filter.label : undefined}
  />
);
