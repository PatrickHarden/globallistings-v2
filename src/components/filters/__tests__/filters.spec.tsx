import React from 'react';
import { shallow } from 'enzyme';
import { FilterType, FilterPlacement, FilterModifier } from '../../../types/config';
import { Filters } from '../filters';
import { SelectFilterInfo, RangeFilterInfo } from '../../../redux/selectors/config/filters';

describe(Filters, () => {
  const props = {
    displayLabels: false,
    handleChange: jest.fn(),
    inline: false,
    filters: [],
  };

  const getSelectFilter = (name: string): SelectFilterInfo => ({
    name,
    label: 'select filter',
    type: FilterType.select,
    items: [{ value: 'foo', label: 'bar', default: false }],
    placement: FilterPlacement.primary,
    modifier: FilterModifier.query,
  });

  const getRangeFilter = (name: string): RangeFilterInfo => ({
    name,
    label: 'select filter',
    type: FilterType.range,
    min: { items: [] },
    max: { items: [] },
    placement: FilterPlacement.secondary,
    modifier: FilterModifier.query,
  });

  it('should render correctly', () => {
    expect(shallow(<Filters {...props} />)).toMatchSnapshot();
  });

  it('should render correctly with filters', () => {
    expect(
      shallow(<Filters {...props} filters={[getSelectFilter('select'), getRangeFilter('range')]} />)
    ).toMatchSnapshot();
  });

  it('should render correctly with filter labels', () => {
    expect(
      shallow(
        <Filters
          {...props}
          displayLabels={true}
          filters={[getSelectFilter('select'), getRangeFilter('range')]}
        />
      )
    ).toMatchSnapshot();
  });

  it('should render correctly with a placement specified', () => {
    expect(
      shallow(
        <Filters
          {...props}
          displayLabels={true}
          filters={[getSelectFilter('select'), getRangeFilter('range')]}
          placement={FilterPlacement.primary}
        />
      )
    ).toMatchSnapshot();
  });
});
