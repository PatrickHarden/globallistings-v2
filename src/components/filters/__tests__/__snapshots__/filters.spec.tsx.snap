// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters should render correctly 1`] = `<styled.div />`;

exports[`Filters should render correctly with a placement specified 1`] = `
<styled.div>
  <SelectFilter
    displayLabel={true}
    filter={
      Object {
        "items": Array [
          Object {
            "default": false,
            "label": "bar",
            "value": "foo",
          },
        ],
        "label": "select filter",
        "modifier": "query",
        "name": "select",
        "placement": "lm_primaryFilter",
        "type": "select",
      }
    }
    inline={false}
    key="select-select"
  />
</styled.div>
`;

exports[`Filters should render correctly with filter labels 1`] = `
<styled.div>
  <SelectFilter
    displayLabel={true}
    filter={
      Object {
        "items": Array [
          Object {
            "default": false,
            "label": "bar",
            "value": "foo",
          },
        ],
        "label": "select filter",
        "modifier": "query",
        "name": "select",
        "placement": "lm_primaryFilter",
        "type": "select",
      }
    }
    inline={false}
    key="select-select"
  />
  <RangeFilter
    displayLabel={true}
    filter={
      Object {
        "label": "select filter",
        "max": Object {
          "items": Array [],
        },
        "min": Object {
          "items": Array [],
        },
        "modifier": "query",
        "name": "range",
        "placement": "lm_secondaryFilter",
        "type": "range",
      }
    }
    handleChange={
      [MockFunction] {
        "calls": Array [
          Array [
            "select",
          ],
          Array [
            "select",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    inline={false}
    key="range-range"
  />
</styled.div>
`;

exports[`Filters should render correctly with filters 1`] = `
<styled.div>
  <SelectFilter
    displayLabel={false}
    filter={
      Object {
        "items": Array [
          Object {
            "default": false,
            "label": "bar",
            "value": "foo",
          },
        ],
        "label": "select filter",
        "modifier": "query",
        "name": "select",
        "placement": "lm_primaryFilter",
        "type": "select",
      }
    }
    inline={false}
    key="select-select"
  />
  <RangeFilter
    displayLabel={false}
    filter={
      Object {
        "label": "select filter",
        "max": Object {
          "items": Array [],
        },
        "min": Object {
          "items": Array [],
        },
        "modifier": "query",
        "name": "range",
        "placement": "lm_secondaryFilter",
        "type": "range",
      }
    }
    handleChange={
      [MockFunction] {
        "calls": Array [
          Array [
            "select",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    inline={false}
    key="range-range"
  />
</styled.div>
`;
